### PR TITLE
Handle nullref when the project is not saved on disk - Closes NuGet/Home#4567

### PR DESCRIPTION
**Pull request by nkolev92**
Closes NuGet/Home#4567

The root cause of our null ref is the fact that the DTE returns an empty solution path when this variable is unchecked.
We make that call right here:
https://github.com/NuGet/NuGet.Client/blob/b89bb5750c475212213e14f750604d2ecaf704ff/src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L531
This however hasn't changed so we need to work around this issue on our side.

The ProjectFactory might return a null NuGetProject, we null check and throw an exception to stop the restore.

Similar to #1205

//cc
@emgarten @zhili1208 @alpaix @rohit21agrawal @mishra14 @jainaashish

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 1]
```diff
@@ -813,8 +813,12 @@ private void AddEnvDTEProjectToCache(EnvDTE.Project envDTEProject)
             ProjectNames oldEnvDTEProjectName;
             _projectSystemCache.TryGetProjectNameByShortName(EnvDTEProjectUtility.GetName(envDTEProject), out oldEnvDTEProjectName);
 
-            // Create the NuGet project first. If this throws we bail out and do not change the cache.
+            // Create the NuGet project first. If this throws or returns null we bail out and do not change the cache.
             var nuGetProject = CreateNuGetProject(envDTEProject);
+            if(nuGetProject == null)
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: nit: space after `if`. Format changed code block.

File: src/NuGet.Clients/PackageManagement.VisualStudio/Strings.resx [hunk 2]
```diff
@@ -236,4 +236,7 @@
   <data name="CredentialProviderFailed_ImportedProvider" xml:space="preserve">
     <value>Failed to load credential provider from assembly {0}.</value>
   </data>
-</root>
+  <data name="Error_FailureToCreateNuGetProject" xml:space="preserve">
+    <value>Cannot create NuGet project. The VS project might be invalid or the solution is not saved on disk.</value>
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/Strings.resx]: not sure about the abbreviation in user message. Replace with full name?

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 3]
```diff
@@ -813,8 +813,12 @@ private void AddEnvDTEProjectToCache(EnvDTE.Project envDTEProject)
...
+            {
+                throw new InvalidOperationException(Strings.Error_FailureToCreateNuGetProject);
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Write error to the activity log?

**Review by alpaix — CHANGES_REQUESTED** [review]
Not sure this change would help user understand how to fix the problem. Let's sync offline.

**Comment by nkolev92** [comment]
:bell: @alpaix

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs [hunk 4]
```diff
@@ -181,6 +181,15 @@ internal sealed class SolutionRestoreJob : ISolutionRestoreJob
             {
                 var solutionDirectory = _solutionManager.SolutionDirectory;
                 var isSolutionAvailable = _solutionManager.IsSolutionAvailable;
+                if (solutionDirectory == null) {
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs]: nit: new line

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 5]
```diff
@@ -813,7 +813,7 @@ private void AddEnvDTEProjectToCache(EnvDTE.Project envDTEProject)
             ProjectNames oldEnvDTEProjectName;
             _projectSystemCache.TryGetProjectNameByShortName(EnvDTEProjectUtility.GetName(envDTEProject), out oldEnvDTEProjectName);
 
-            // Create the NuGet project first. If this throws we bail out and do not change the cache.
+            // Create the NuGet project first. If this throws or returns null we bail out and do not change the cache.
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: not relevant after you unrolled your changes here.

**Review by alpaix — APPROVED** [review]
Couple of minor notes. After that 🚢.

**Comment by nkolev92** [comment]
Can one more person take a look?

@emgarten @rohit21agrawal @jainaashish @zhili1208 @mishra14

**Review by emgarten — APPROVED** [review]

**Pull request merged**